File: Dreya/torrent/MkvFile.java
package dreya.torrent;

import dreya.Config;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import dreya.messenger.Sender;

/**
 *
 * @author smv
 */
public class MkvFile {
    private String filename;

    private int subtitleMaxScore = 0;
    private String subtitleUrl = "";
    
    private int season = 0;
    private int episode = 0;
    
    private String s_season = "";
    private String s_episode = "";
    
    
    public MkvFile(String filename)
    {
        this.filename = filename;
        
        Pattern MY_PATTERN = Pattern.compile("S[0-9].E[0-9].");
        Matcher m = MY_PATTERN.matcher(filename);
        
        if (m.find())
        {
            String seepcode = m.group();
            seepcode = seepcode.toLowerCase();
            seepcode = seepcode.replace("s","");
            
            String[] parts = seepcode.split("e");
            
            this.season = Integer.parseInt(parts[0]);
            this.episode = Integer.parseInt(parts[1]);
        }
    }
    
    public boolean isValidWithoutSrt()
    {
        if (this.filename.length()>3)
        {
            
            if ("mkv".equals(this.filename.substring(this.filename.length() - 3)))
            {
                
                String srtFile = this.getSrtFileName();
                
                File f = new File(Config.showsFolder+srtFile);
                return !(f.exists() && !f.isDirectory());
            }
        }
        
        return false;
    }
    
    private String getSrtFileName()
    {
        return this.filename.substring(0,this.filename.length()-3)+"srt";
    }
    
    public void checkSubtitle(String content, String href)
    {
        
        String filename = this.filename.toLowerCase();
        content = content.toLowerCase();
        
        String s_ep = this.episode+"";        
        String d_ep = this.episode+"";
        if (this.episode<10)
        {
            d_ep = "0"+this.episode;
        }
        this.s_episode = d_ep;
        String s_se = this.season+"";        
        String d_se = this.season+"";
        if (this.season<10)
        {
            d_se = "0"+this.season;
        }
        this.s_season = d_se;
        
        if (content.contains("magyar"))
        {
            content = content.replace("[", " ");
            content = content.replace("]", " ");
            content = content.replace("(", " ");
            content = content.replace(")", " ");            
            content = content.replace(".", " ");
            content = content.replace(",", " ");
            content = content.replace("-", " ");
            
            while (!content.equals(content.replace("  "," ")))
            {
                content = content.replace("  ", " ");    
            }
            
           
            if (content.contains(" "+s_se+"x"+d_ep+" "))
            {
                int score = 0;
                String[] parts = content.split(" ");
                 
                for (String part: parts)
                {
                    if (filename.contains(part))
                    {
                         score++;
                    }
                }
                
                if (score>this.subtitleMaxScore)
                {
                    this.subtitleMaxScore = score;
                    this.subtitleUrl = href;
                }
            }
            
            
            
           
        }
    }
    
    public void download()
    {
        boolean hasSeries = false;
        
        if (this.subtitleMaxScore>0)
        {
            try
            {
                
                List<Series> seriesList = Config.seriesList;
              
                for (Series series : seriesList)
                {
                    if (this.filename.toLowerCase().contains(series.getSubtitleText().toLowerCase()))
                    {
                        
                        Sender.send("", series.getName()+" "+this.s_season+"x"+this.s_episode+" magyar felirat letöltve!");
                        hasSeries = true;
                    }
                }
                
                if (!hasSeries)
                {
                    Sender.send("", this.getSrtFileName()+" magyar felirat letöltve!");
                }
                
                System.setProperty("http.agent", "Mozilla/5.0 (Android 4.4; Mobile; rv:41.0) Gecko/41.0 Firefox/41.0"); 

                URL website = new URL("http://feliratok.info"+this.subtitleUrl);
                ReadableByteChannel rbc = Channels.newChannel(website.openStream());
                FileOutputStream fos = new FileOutputStream(Config.showsFolder+this.getSrtFileName());
                fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);                        
            }
            catch (IOException ex) {
                System.out.println("[error] "+ex.getMessage());
            }
        }
    }
    
}

File: Dreya/Series/MkvFile.py
# -*- coding: utf-8 -*-

import os.path
import re
from Messenger.Messenger import Messenger
from Http import Http
from Terminal import Terminal

class MkvFile:
    file = ""
    parts = ""
    series = ""
    maxurl = ""
    maxscore = 0
    serep = ""

    def __init__(self, file):
        self.file = file
        self.parts = self.normalizer(file)


    def normalizer(self, text):
        text = text.lower()
        text = text.replace("."," ")
        text = text.replace("-", " ")
        text = text.replace("]", " ")
        text = text.replace("[", " ")
        text = text.replace(")", " ")
        text = text.replace("(", " ")
        text = text.replace(":", " ")
        text = text.replace("!", " ")
        text = text.replace("_", " ")
        text = text.replace("/", " ")
        text = text.replace(",", " ")
        text = text.replace("  ", " ")
        text = text.replace("  ", " ")
        text = text.replace("  ", " ")

        return text

    def isSearcheable(self):
        from Config import Config

        for series in Config.series:
            if series.checkSubtitle(self.parts):
                if not os.path.isfile(self.file.replace(".mkv", ".srt")):
                    self.series = series
                    return True

        return False

    def downloadMaxScore(self):
        if self.maxscore > 0:
            try:
                Http.download("http://feliratok.info"+self.maxurl, self.file.replace(".mkv", ".srt"))
                Messenger.sendMessage("",self.series.name + self.serep + " magyar felirat letoltve!")
                Terminal.info(self.series.name + self.serep + " magyar felirat letoltve!", "info")
            except Exception as ex:
                print(ex)
                return 0

    def subtitle(self, text,  url):
        check = self.normalizer(text)

        try:
            output = re.search('s[0-9].e[0-9].', check, flags=re.IGNORECASE)
            if output:
                output = output.group(0)
                ep = output.split("e")
            else:
                output = re.search('[0-9]*x[0-9]*', check, flags=re.IGNORECASE)
                if output:
                    output = output.group(0)
                    ep = output.split("x")
                else:
                    return 0

            if len(ep) > 1:
                season = int(ep[0])
                origseason = str(int(ep[0]))
                episode = int(ep[1])

                if episode < 10:
                    episode = "0" + str(episode)

                episode = str(episode)

                if season < 10:
                    season = "0" + str(season)

                season = str(season)

                find1 = "s" + season + "e" + episode
                find2 = season + "x" + episode
                find3 = "s" + origseason + "e" + episode
                find4 = origseason + "x" + episode

                if find1 in self.parts or find2 in self.parts or find3 in self.parts or find4 in self.parts:
                    if self.series.checkSubtitle(check):
                        self.serep = find1.upper()
                        subtitleparts = check.split(" ")

                        score = 0
                        for part in subtitleparts:
                            if part in self.file:
                                score = score + 1

                        if score > self.maxscore:
                            self.maxscore = score
                            self.maxurl = url

        except Exception as ex:
            return 0
